### DISCUSSION TITLE
Remove Bg

### DISCUSSION THREAD

**Discussion by AceroMagico** [WASasquatch/was-node-suite-comfyui#23] [Ideas]
It would be great if you could integrate the remove background (Also very useful for Depth maps), as done in automatic1111 : 

https://github.com/thygate/stable-diffusion-webui-depthmap-script
https://github.com/danielgatis/rembg 

It's not so good as clipdrop https://clipdrop.co/
but is not bad at all and it's free, here a comparison : https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/8541

About this comparison, however, it must be said that I think clipdrop does an upscale before performing the key cut and then have a cleaner key.
I tried doing this in auto111 and actually then remove bg cuts almost as well as clipdrop.

( I warn you, however, that installing the required onnxruntime-gpu library is a bit of a mess, in auto111 win10 for to make it work I had to do some wizardry: The problem is that by doing that the other linked libraries go out of version. But once installed you can pip install --force-reinstall --no-deps of the others one at a time. while if you don't use the --no-deps each library brings back the linked versions and the onnxruntime-gpu doesn't work (loop) )

**Comment by WASasquatch** [comment]
I am not using any libraries that have any sort of compatibility issues that could make the automatic installation process require any manual work. 

Also they using an API to remove backgrounds, not their own code. According to their website the removed background images are for attribution use unless you are paying for the service. That's no good for a MIT commercial project.

**Comment by AceroMagico** [comment]
Regarding the onnxruntime-gpu library, I understand your point of view. In fact, I wanted to warn you, although it must be said that it is not the library itself that is having problems.

As for the removebg itself, I don't think you read it right or I apologize for not being clear myself:

Clipdrop, which as you say uses api and is not free, I only pointed it out to you as a comparison.

The subject of the message is rembg
https://github.com/danielgatis/rembg which is instead free Licensed under [MIT License](https://github.com/danielgatis/rembg/blob/main/LICENSE.txt)

**Reply by WASasquatch** [reply]
> The subject of the message is rembg https://github.com/danielgatis/rembg which is instead free Licensed under [MIT License](https://github.com/danielgatis/rembg/blob/main/LICENSE.txt)

Ah, I see what you mean. onnxruntime-gpu appears to have too many issues to consider this right now. I followed directions to a T, CUDA 11.8 for my 4090 and everything but couldn't get it to install on my system. Even used a fresh conda so no conflicts with anything else I have to test. So I can't even test this to add it.

Also, they don't allow you to set the model location which is odd. I don't want stuff in python home directory that isn't where I manage my files (such as ComfyUI models folders). Especially for large files you may forget about and wonder where space is going. 

I feel ONNX still has a ways to mature. I've seen so many issues with it down to inconsistent benchmark timings from run to run.

**Comment by AceroMagico** [comment]
Definitely install it is a mess, I also have CUDA 11.8 for my 4090, in automatic1111 I was able to do it with the system I mentioned above (once installed you can pip install --force-reinstall --no-deps of the others one at a time) and it works very well both in the extra tab and with the extension https://github.com/thygate/stable-diffusion-webui-depthmap-script.

I understand that you don't want to have files outside of your parent folder.

**Reply by WASasquatch** [reply]
I wonder if it's possible to use something like: https://github.com/ENOT-AutoDL/onnx2torch and convert the models, and use them with PyTorch (which would already be installed for ComfUI), and just host the models on HF or something.  Not sure about the usage though once converted x.x. lol

**Comment by AceroMagico** [comment]
To be honest, I see it much easier to have the files out of your parent folder than to venture into a conversion and host.

**Reply by WASasquatch** [reply]
> To be honest, I see it much easier to have the files out of your parent folder than to venture into a conversion and host.

It's not good to install large files to random caches the user may not even know where it is, or forget about. Especially on windows where you can't simply stat a HDD for the largest files easily. Additionally, it's not just about just having it stored somewhere else, but all the inherent problems with ONNX system across installs. I am not going to deal with people not being able to install it when I can't even figure it out. Lol

**Comment by AceroMagico** [comment]
I agree with you, at present it is not advisable to attempt any of the ways analyzed. 
I have noticed how people do not realize how enjoyable but also complex it is to program and turn to developers of free applications as if they had paid for them. 
I do not like this at all.

Maybe when ONNX is more integrated we can reconsider. ;)

**Comment by AceroMagico** [comment]
May be there's another way:

demo: https://segment-anything.com/dataset/index.html

https://segment-anything.com/

In this landing website they say that : The prompt encoder and mask decoder `can run directly with PyTorch ` or converted to ONNX and run efficiently on CPU or GPU across a variety of platforms that support ONNX runtime.

even if then in github there is:
The following optional dependencies are `necessary for mask post-processing`, saving masks in COCO format, the example notebooks, and exporting the model in ONNX format. jupyter is also required to run the example notebooks.

pip install opencv-python pycocotools matplotlib onnxruntime onnx

https://github.com/facebookresearch/segment-anything

(Off Topic: The implications of this type of technology applied to drones moved by an AI can be  very worrying):

**Comment by WASasquatch** [comment]
Pycocotools needs special build tools for C++ so won't rely on anything with pycoco. Almost dumped BLIP for same reason. I can't even install build tools it needs cause I used 2020 version for other software, and it won't build the package. So anything that requires users to download specific versions that make other stuff of theirs incompatible I won't be using.
